src/main.cpp

```diff
@@ -2177,75 +2383,153 @@ bool DisconnectBlock(CBlock& block, CValidationState& state, CBlockIndex* pindex
+        return error("DisconnectBlock(): failure reading undo data");
+
+    LogPrint("sc", "%s():%d - ===============> CBlockUndo red from DB:\n%s\n",
+        __func__, __LINE__, blockUndo.ToString());
+    // no coinbase in blockundo
+#if 0
+    if (blockUndo.vtxundo.size() + 1 != block.vtx.size() + block.vcert.size() )
+#else
+    if (blockUndo.vtxundo.size() < (block.vtx.size() - 1 + block.vcert.size()))
+#endif
+        return error("DisconnectBlock(): block and undo data inconsistent");
+
+    LogPrint("sc", "%s():%d - restoring sc coins if any\n", __func__, __LINE__);
+    if (!view.RestoreImmatureBalances(pindex->nHeight, blockUndo) )
+    {
+        LogPrint("sc", "%s():%d - ERROR updating sc maturity amounts\n", __func__, __LINE__);
+        return error("DisconnectBlock(): sc and undo data inconsistent");
+    }
+
+    for(int idx = block.vtx.size() - 1 + block.vcert.size(); idx < blockUndo.vtxundo.size(); idx++)
+    {
+        LogPrint("sc", "%s():%d - calling RevertCeasingScs idx[%d]\n", __func__, __LINE__, idx);
+        bool ret = view.RevertCeasingScs(blockUndo.vtxundo[idx]);
+        if (!ret)
+            return error("DisconnectBlock(): cannot revert ceasing sc");
+    }
+
+    // not including coinbase
+    const int certOffset = block.vtx.size() - 1;
+
+    // undo certificates in reverse order
+    for (int i = block.vcert.size() - 1; i >= 0; i--) {
+        const CScCertificate& cert = block.vcert[i];
+        uint256 hash = cert.GetHash();
+
+        LogPrint("cert", "%s():%d - reverting outs of cert[%s]\n", __func__, __LINE__, hash.ToString());
+
+        // Check that all outputs are available and match the outputs in the block itself
+        // exactly.
+        {
+            CCoinsModifier outs = view.ModifyCoins(hash);
+            outs->ClearUnspendable();
+ 
+            CCoins outsBlock(cert, pindex->nHeight);
+            // The CCoins serialization does not serialize negative numbers.
+            // No network rules currently depend on the version here, so an inconsistency is harmless
+            // but it must be corrected before txout nversion ever influences a network rule.
+            if (outsBlock.nVersion < 0)
+                outs->nVersion = outsBlock.nVersion;
+
+
+            if (*outs != outsBlock) {
+
+                LogPrint("cert", "%s():%d - outs     :%s\n", __func__, __LINE__, outs->ToString());
+                LogPrint("cert", "%s():%d - outsBlock:%s\n", __func__, __LINE__, outsBlock.ToString());
+
+                fClean = fClean && error("DisconnectBlock(): added transaction mismatch? database corrupted");
+                LogPrint("cert", "%s():%d - tx[%s]\n", __func__, __LINE__, hash.ToString());
+            }
+  
+            // remove outputs
+            LogPrint("cert", "%s():%d - clearing outs of tx[%s]\n", __func__, __LINE__, hash.ToString());
+            outs->Clear();
+        }
+
+        if (!view.UndoCeasingScs(cert)) {
```

Comment:
aggregate UndoCeasingScs and RevertCertOutputs in a new method RevertFromScInfo  taking as parameter a CCertificate